app-experimental.tex

```diff
@@ -1311,6 +1315,69 @@ \subsection{Points-to-PCC} % <<<
+This code capability cannot be used in regular jumps because it is untagged.
+The only way it can be used for a jump is in a subsequent cjaurl instruction.
+
+An attacker can only jump to a code capability with the isentry it came from.
+This can be divided into two cases: If the instruction following the lil is a cjaurl, this will jump to the code pointed to by the code capability and unseal the isentry.
+The register holding the isentry is stored in the reservation information of the lil instruction.
+In the second case, if the instruction following the lil is not a cjaurl, the reservation will have become invalid.
+Thus the attacker cannot jump to the code capability.
+Equally, the isentry has never been unsealed and thus cannot be used by the attacker.
+
+Because of this linkage, the attacker can never use a code capability and a data capability together, which have not been intended to work together.
+Thus the linking keeps the atomicity of the instruction.
+
+Loading the code capability does not break CHERI confidentiality.
+Isentries are not designed for secrecy about where a compartment wants to jump, but to implement jumping to a sealed pair of capabilities.
+Similar methods, e.g., cinvoke gives out sealed code and data capabilities, which are then invoked as a pair.
+The code capability itself is not secret.
+
+It is important to note that the linked pair of a lil and cjaurl instruction does not give any guarantees about memory consistency. Memory consistency is not an issue tackled by CHERI, but needs to be addressed by other architectural mechanisms, e.g. atomic memory operations.
+
+\subsubsection{Allowing more instructions in between lil and cjaurl}
+
+In order to enhance performance, we want to allow for ordinary instructions to appear between a lil and cjaurl instruction.
+This could allow for the load-to-use penalty to be reduced.
+
+The invariants laid out above still need to hold: One must only jump to a linked code capability with a cjaurl instruction.
+The linked code capability can only be used with the isentry it came from and the other way around.
+
+We envision the following way of implementing it.
+The lil instruction makes a reservation, which stores the following information: a one bit tag whether the reservation is valid, a five bit index to indicate the register of the isentry.
+A write to the register of the isentry linked will clear the valid bit of the reservation.
+A write to ct6 -- where the code capability is stored -- will equally lead to the reservation being invalidated.
+
+There always is only ever one reservation.
+A lil instruction is a write to the ct6 register and thus invalidates the reservation.
+A read of any register does not invalidate the reservation.
+None of the information is secret and the reservation will still point to a correct copy of the isentry as well as the code capability.
+
+In this proposal there is no maximum number of instructions between the lil and cjaurl instruction.
+
+
```

Comment:
What happens if the core takes an interrupt between `lil` and `cjaurl`?  Is it possible to spill and restore these reservations?  (That is, how does context switching work?)

Reply:
Or you take a page fault (whether because you put a memory instruction in between lil and cjaurl, or because those two instructions straddled a code page boundary)?

Reply:
(and I wouldn't be thrilled about the OS being able to forge such reservations... it effectively becomes a somewhat-janky capability forgery oracle)

Reply:
Ah, when we discussed this, I think the idea was that it can fail. You have to put it in a loop like LR SC.

Reply:
Then you can't have an arbitrary, or even just unbounded, set of instructions in between, because you need a forward progress guarantee

Reply:
Yep, we discussed the pattern being roughly: LIL, busy work (presumably register clearing), CJAURL, LIL, CJAURL, unconditional jump back 2. The first two are optimistic, but the last 2 are the only ones that would need a progress guarantee.

Reply:
I should have phrased it differently. I was aware of the need for forward progress guarantee, but I am not yet sure what a sensible number of instructions is. I will have a think about what a good number is. Having a forward progress guarantee only if the cjaurl follows a lil instruction could incentivise microarchitects to optimise for that which would harm the case of pushing the lil before doing work and thus attempting to mitigate a potential load to use penalty.